tests/subsys/net/lib/nrf_cloud/cloud/src/main.c

```diff
@@ -0,0 +1,1880 @@
+
+	/* Message data */
+	struct nrf_cloud_tx_data tx_data_t = {
+		.topic_type = NRF_CLOUD_TOPIC_BULK,
+		.qos = MQTT_QOS_0_AT_MOST_ONCE
+	};
+
+	int ret = nrf_cloud_send(&tx_data_t);
+
+	zassert_ok(ret, "return should be 0 when success");
+}
+
+/* Verify that the function will fail if an invalid topic_type is given */
+ZTEST(nrf_cloud_test, test_cloud_send_invalid_topic_type)
+{
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	/* Message data */
+	struct nrf_cloud_tx_data tx_data_t = {
+		.topic_type = 0,
+		.qos = MQTT_QOS_0_AT_MOST_ONCE
+	};
+
+	int ret = nrf_cloud_send(&tx_data_t);
+
+	zassert_equal(-ENODATA, ret, "return should be -ENODATA with invalid topic type");
+}
+
+#if defined(CONFIG_NRF_CLOUD_CONNECTION_POLL_THREAD)
+
+/* TEST START_CONNECTION_POLL */
+
+/* Verify that the function will fail
+ * if the cloud is in the idle state
+ */
+ZTEST(nrf_cloud_test, test_start_connection_poll_in_idle)
+{
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	int ret = start_connection_poll();
+
+	zassert_equal(-EACCES, ret, "return should be -EACCES when in idle state");
+}
+
+/* Verify that the function will fail
+ * if the poll thread is already active
+ */
+ZTEST(nrf_cloud_test, test_start_connection_poll_already_active)
+{
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	/* Init the cloud successfully */
+	init_cloud_success();
+
+	zassert_equal(STATE_INITIALIZED, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the initialized state after initialization");
+	(void)start_connection_poll();
+
+	int ret = start_connection_poll();
+
+	zassert_equal(-EINPROGRESS, ret, "return should be -EINPROGRESS when already active");
+}
+
+/* Verify that the function will succeed
+ * if the cloud is successfully initialized
+ */
+ZTEST(nrf_cloud_test, test_start_connection_poll_success)
+{
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	/* Init the cloud successfully */
+	init_cloud_success();
+
+	int ret = start_connection_poll();
+
+	zassert_ok(ret, "return should be 0 when success");
+}
+
+/* TEST NRF_CLOUD_RUN */
+
+/* Verify that the function will change the cloud disconnect
+ * event to not success if the connection to the cloud failed
+ */
+ZTEST(nrf_cloud_test, test_cloud_run_cloud_connecting_failed)
+{
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	/* Init the cloud successfully */
+	init_cloud_success();
+
+	/* Connect to the cloud failed */
+	nct_connect_fake.custom_fake = fake_nct_connect__invalid;
+
+	(void)start_connection_poll();
```

Comment:
You're calling start_connection_poll but the test refers to NRF_CLOUD_RUN, is that intentional?

Reply:
It's intentional, since `start_connection_poll` will trigger the thread which causes `nrf_cloud_run` actually run.

Reply:
Ah, yeah, I actually thought this was one of the NRF_CLOUD_RUN tests, not one of the CONNECTION_POLL_THREAD tests, but I see now I had that wrong -- I guess, further evidence that the tests should refer to the function they are testing rather than "the function" in the description 😅

Sorry about the confusion

Reply:
I misremembered the purpose of nrf_cloud_run, my bad